### PR TITLE
Validation of requests in garbage-collect

### PR DESCRIPTION
**Pull request by bartsmykla**
Closes https://github.com/openfaas/openfaas-cloud/issues/140

Function git-tar is now signing requests to garbage-collect
using payload-secret and garbage-collefct function is validating
request before starting garbage collecting.

I have modified .gitignore to not commit by mistake basic-auth-user
and basic-auth-password file

Signed-off-by: Bart Smykla (VMware) <bartek@smykla.com>

## How Has This Been Tested?
I have send unsigned and signed requests (with valid payloads and with some random ones).

## Checklist:

I have:

- [x] updated the documentation if required
- [x] read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] signed-off my commits with `git commit -s`
- [ ] added unit tests

### CODE REVIEW

File: garbage-collect/handler.go [hunk 1]
```diff
@@ -11,14 +11,21 @@ import (
...
+	"github.com/alexellis/hmac"
 	"github.com/openfaas/openfaas-cloud/sdk"
 )
 
 // Handle a serverless request
 func Handle(req []byte) string {
+	err := validateRequest(req)
+
+	if err != nil {
+		log.Panic(err)
```
Comment (alexellis) [inline garbage-collect/handler.go]: Are we doing `log.Panic` elsewhere in this scenario? Check `github-push` for instance or one of the others.

File: garbage-collect/handler.go [hunk 2]
```diff
@@ -11,14 +11,21 @@ import (
...
+	}
 
 	garbageReq := GarbageRequest{}
-	err := json.Unmarshal(req, &garbageReq)
+	err = json.Unmarshal(req, &garbageReq)
```
Comment (alexellis) [inline garbage-collect/handler.go]: Please use a different name for the first `err` or the second one.

File: garbage-collect/handler.go [hunk 3]
```diff
@@ -48,6 +55,28 @@ func Handle(req []byte) string {
...
+func validateRequest(req []byte) (err error) {
+	payloadSecret, err := sdk.ReadSecret("payload-secret")
+
+	if err != nil {
+		return fmt.Errorf("couldn't get payload-secret: %t", err)
+	}
+
+	xCloudSignature := os.Getenv("Http_X_Cloud_Signature")
+
+	if xCloudSignature == "" {
```
Comment (alexellis) [inline garbage-collect/handler.go]: Not sure we need this check. It will be covered by the following statement.

File: garbage-collect/handler.go [hunk 4]
```diff
@@ -48,6 +55,28 @@ func Handle(req []byte) string {
...
+		return fmt.Errorf("unauthorized request")
+	}
+
+	err = hmac.Validate(req, xCloudSignature, payloadSecret)
+
+	if err != nil {
+		return fmt.Errorf("unauthorized request")
```
Comment (alexellis) [inline garbage-collect/handler.go]: What error message are we using elsewhere when the digest doesn't match or is invalid? We should aim to be consistent across the board to help with maintenance.

File: git-tar/function/handler.go [hunk 5]
```diff
@@ -118,8 +120,19 @@ func collect(pushEvent sdk.PushEvent, stack *stack.Services) error {
...
+
+	payloadSecret, err := getPayloadSecret()
+
+	if err != nil {
+		return fmt.Errorf("failed to load payload secret, error %t", err)
+	}
+
 	request, _ := http.NewRequest(http.MethodPost, gatewayURL+"function/garbage-collect", bufferReader)
 
+	digest := hmac.Sign(bytesReq, []byte(payloadSecret))
```
Comment (alexellis) [inline git-tar/function/handler.go]: Looks fine on this side.

**Comment by bartsmykla** [comment]
@alexellis all should be fixed now :-)

**Comment by alexellis** [comment]
Did you test this before/after the latest change? Can you test again and share the input/output from your curl commands?

**Comment by bartsmykla** [comment]
@alexellis I have tested it again and it's working as expected! :-)

**Comment by alexellis** [comment]
@bartsmykla 

> Can you test again and share the input/output from your curl commands?

**Pull request merged**